File: solutions/monotonic-stack/084-largest-rectangle-in-histogram.py
"""
# 084. Largest Rectangle In Histogram
# Difficulty: Medium
Given a problem that demonstrates key concepts in Monotonic Stack.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of monotonic stack concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply monotonic stack methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages monotonic stack principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses monotonic stack techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using monotonic stack method
3. Return the computed result

</details>
"""

class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        """
        Find largest rectangle area in histogram using monotonic stack.

        Args:
            heights: Array of bar heights in histogram

        Returns:
            Maximum rectangle area

        Time Complexity: O(n) - each element pushed/popped once
        Space Complexity: O(n) - stack can hold all elements in worst case
        """
        stack = []  # Store indices of bars
        max_area = 0

        for i, h in enumerate(heights):
            # Maintain increasing stack
            while stack and heights[stack[-1]] > h:
                height = heights[stack.pop()]
                # Width is distance to current bar minus distance to previous bar
                width = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, height * width)

            stack.append(i)

        # Process remaining bars in stack
        while stack:
            height = heights[stack.pop()]
            width = len(heights) if not stack else len(heights) - stack[-1] - 1
            max_area = max(max_area, height * width)

        return max_area

    def solve(self, heights: list[int]) -> int:
        """Wrapper method for consistency with template."""
        return self.largestRectangleArea(heights)

def test_solution():
    """
    Test cases for 084. Largest Rectangle In Histogram.
    """
    solution = Solution()

    # Test case 1: Classic example
    result = solution.solve([2, 1, 5, 6, 2, 3])
    expected = 10
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Single bar
    result = solution.solve([2])
    expected = 2
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: Increasing heights
    result = solution.solve([1, 2, 3, 4, 5])
    expected = 9
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 4: Decreasing heights
    result = solution.solve([5, 4, 3, 2, 1])
    expected = 9
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 5: All same height
    result = solution.solve([4, 4, 4, 4])
    expected = 16
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 6: Empty array
    result = solution.solve([])
    expected = 0
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 7: Two bars
    result = solution.solve([2, 4])
    expected = 4
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    heights = [2, 1, 5, 6, 2, 3]
    result = solution.solve(heights)
    print(f"Solution for 084. Largest Rectangle In Histogram: {result}")

File: solutions/stack/alternatives/084-largest-rectangle-in-histogram.js
/**
 * 84. Largest Rectangle In Histogram
 * Medium
 *
 * Largest Rectangle in Histogram Time Complexity: O(n) Space Complexity: O(n) @param {number[]} heights - Array of integers representing histogram bar heights @return {number} - The area of the largest rectangle
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Largest Rectangle In Histogram is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Largest Rectangle in Histogram
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 * 
 * @param {number[]} heights - Array of integers representing histogram bar heights
 * @return {number} - The area of the largest rectangle
 */
function largestRectangleArea(heights) {
    if (!heights || heights.length === 0) return 0;
    
    const stack = [];
    let maxArea = 0;
    let i = 0;
    
    // Process all bars in the histogram
    while (i < heights.length) {
        // If stack is empty or current height is larger than previous,
        // push current index to stack
        if (stack.length === 0 || heights[stack[stack.length - 1]] <= heights[i]) {
            stack.push(i);
            i++;
        } else {
            // Calculate area with the top bar as smallest bar
            maxArea = calculateArea(heights, stack, maxArea, i);
        }
    }
    
    // Process remaining bars in the stack
    while (stack.length > 0) {
        maxArea = calculateArea(heights, stack, maxArea, i);
    }
    
    return maxArea;
}

/**
 * Helper function to calculate area for current bar
 * 
 * @param {number[]} heights - Array of heights
 * @param {number[]} stack - Stack of indices
 * @param {number} maxArea - Current maximum area
 * @param {number} i - Current index
 * @return {number} - Updated maximum area
 */
function calculateArea(heights, stack, maxArea, i) {
    const height = heights[stack.pop()];
    const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
    return Math.max(maxArea, height * width);
}

/**
 * Test cases
 */
function runTests() {
    const testCases = [
        {
            input: [2, 1, 5, 6, 2, 3],
            expected: 10,
            description: "Regular case with varying heights"
        },
        {
            input: [2, 4],
            expected: 4,
            description: "Small array"
        },
        {
            input: [],
            expected: 0,
            description: "Empty array"
        },
        {
            input: [1],
            expected: 1,
            description: "Single element"
        },
        {
            input: [2, 2, 2, 2],
            expected: 8,
            description: "Equal heights"
        }
    ];

    testCases.forEach((testCase, index) => {
        const result = largestRectangleArea(testCase.input);
        console.log(`Test ${index + 1}: ${testCase.description}`);
        console.log(`Input: [${testCase.input}]`);
        console.log(`Expected: ${testCase.expected}`);
        console.log(`Got: ${result}`);
        console.log(`Status: ${result === testCase.expected ? 'PASSED' : 'FAILED'}`);
        console.log('---');
    });
}

// Export the function for use in other modules
module.exports = {
    largestRectangleArea
};

// Run tests if this file is being run directly
if (require.main === module) {
    runTests();
}